[logging] Save compile state in CompiledFxGraph and make it available at runtime

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146455

Summary: To support logging the correct compile_id for runtime timings (like Triton autotuning), save the compile_id in CompiledFxGraph make it available to logging utilities, i.e., dynamo_timed.

The previous attempt put the compile_id in the inductor_metadata with the Triton output code, but that broke Triton caching and we reverted. This version does the following:
* When creating or deserializing a CompiledFxGraph, save the compile-time compile_id.
* Implement a class `RuntimeCompileContext` that's analogous to `CompileContext` where we can look up the compile_id at runtime.
* Set this runtime compile context during `CompiledFxGraph.__call__`.
* Removes the compile_id as a param to dynamo_timed; dynamo_timed can figure it out instead.
* Removes separate dynamo_timed params for compile-time and runtime dynamo_compile column names. We can use one param have dynamo_timed figure out whether to treat as a runtime or compile-time event.

Test Plan:
* tlparse (`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`): https://fburl.com/bu5i8efk
* dynamo_compile: https://fburl.com/scuba/dynamo_compile/sandbox/3d74ps92
* pt2_compile_events: https://fburl.com/scuba/pt2_compile_events/ooqoe5tu

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

masnesral committed:
[logging] Save compile state in CompiledFxGraph to make it available at runtime

Summary: To support logging the correct compile_id for runtime timings (like Triton autotuning), save the compile_id in CompiledFxGraph make it available to logging utilities, i.e., dynamo_timed.

Test Plan:

[ghstack-poisoned]

masnesral committed:
Update on "[logging] Save compile state in CompiledFxGraph to make it available at runtime"


Summary: To support logging the correct compile_id for runtime timings (like Triton autotuning), save the compile_id in CompiledFxGraph make it available to logging utilities, i.e., dynamo_timed.

Test Plan:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

masnesral commented on torch/_dynamo/metrics_context.py:
```diff
@@ -152,16 +150,13 @@ def increment(
             self._metrics[metric] = 0
         self._metrics[metric] += value
 
-        if extra:
```
This was all to thread through a compile_id for runtime events. Not needed any more.

masnesral commented on torch/_dynamo/metrics_context.py:
```diff
@@ -152,16 +150,13 @@ def increment(
         """
         if self._metrics:
             end_time_ns = time.time_ns()
+            if extra:
```
Any final context at the time of logging a runtime record. All that I have now is `is_forward`

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -573,9 +573,6 @@ def dynamo_timed(
     log_pt2_compile_event: bool = False,
     metadata: Optional[dict[str, object]] = None,
     dynamo_compile_column_us: Optional[str] = None,
-    dynamo_compile_runtime_column_us: Optional[str] = None,
```
No longer passed explicitly from the inductor_metadata.

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -611,23 +608,12 @@ def _foo(...):
         assert dynamo_compile_column_us.endswith("_us")
 
-    # Only one of these should be set.
-    assert dynamo_compile_column_us is None or dynamo_compile_runtime_column_us is None
```
I don't need a separate param for runtime vs. compile-time column names. We can figure it out automatically

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -643,13 +629,11 @@ def _foo(...):
         event_metadata.update(metadata)
     if fn_name:
         event_metadata.update({"fn_name": fn_name})
-    if is_forward is not None:
```
This is one thing I removed. It was used to log an "is_backward" entry to the chromium event. Problem is that it was used inconsistently and only provided for the runtime autotuning event. If we need this, I'd prefer to make it work for all dynamo_timed calls (in a follow-up pr)

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1258,8 +1237,6 @@ def safe_str(item: Any) -> str:
         all_metrics["inductor_fx_remote_cache_miss_keys"] = collection_to_str(
             all_metrics.get("inductor_fx_remote_cache_miss_keys")
         )
-        compile_id = all_metrics.get("compile_id")
```
This was here to make sure the compile_id was logged as a string. Not needed because that's done elsewhere.

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1320,6 +1297,60 @@ def add_compilation_metrics_to_chromium(c: CompilationMetrics) -> None:
     )
 
 
+class RuntimeCompileContext:
```
This essentially copies `torch._guards.CompileContext`, but is for providing the compile_id at runtime. I didn't think it belonged in torch._guards since it's just for logging.

masnesral commented on torch/_dynamo/utils.py:
```diff
@@ -1426,10 +1455,6 @@ def record_compilation_metrics(
         # without making it inconsistent with compilation metrics itself, so
         # we ignore the (hopefully small) time spent logging compilation metrics
         record_logging_overhead=False,
-        # These may be runtime logs, e.g., runtime autotunning, so we provide
```
The callee can use the new utilities to get the correct compile_id; don't need to pass it

masnesral commented on torch/_guards.py:
```diff
@@ -93,28 +93,6 @@ def __str__(self):
             assert self.frame_id is not None and self.frame_compile_id is not None
             return f"{self.frame_id}/{self.frame_compile_id}"
 
-    @classmethod
```
This should have been removed when we reverted the previous functionality. We don't need to create CompileContext objects from the string passed in the inductor_metadata any more.

masnesral replied:
@xmfan, wdyt? Do you prefer to keep it around even though we don't need it now?

xmfan replied:
ideally we don't need this in core, but this can help as a test util

masnesral replied:
Ok I'll put it back then

masnesral commented on torch/_inductor/output_code.py:
```diff
@@ -451,16 +457,24 @@ def __init__(
-        finally:
-            get_runtime_metrics_context().finish()
-            AutotuneCacheBundler.end_compile()
+        with runtime_compile_context(self._runtime_ctx):
```
Here's the important part lol

masnesral committed:
Update on "[logging] Save compile state in CompiledFxGraph and make it available at runtime"


Summary: To support logging the correct compile_id for runtime timings (like Triton autotuning), save the compile_id in CompiledFxGraph make it available to logging utilities, i.e., dynamo_timed.

The previous attempt put the compile_id in the inductor_metadata with the Triton output code, but that broke Triton caching and we reverted. This version does the following:
* When creating or deserializing a CompiledFxGraph, save the compile-time compile_id.
* Implement a class `RuntimeCompileContext` that's analogous to `CompileContext` where we can look up the compile_id at runtime.
* Set this runtime compile context during `CompiledFxGraph.__call__`.
* Removes the compile_id as a param to dynamo_timed; dynamo_timed can figure it out instead.
* Removes separate dynamo_timed params for compile-time and runtime dynamo_compile column names. We can use one param have dynamo_timed figure out whether to treat as a runtime or compile-time event.

Test Plan:
* tlparse (`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`): https://fburl.com/bu5i8efk
* dynamo_compile: https://fburl.com/scuba/dynamo_compile/sandbox/3d74ps92
* pt2_compile_events: https://fburl.com/scuba/pt2_compile_events/ooqoe5tu

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

xmfan commented on torch/_dynamo/utils.py:
```diff
@@ -1320,6 +1297,60 @@ def add_compilation_metrics_to_chromium(c: CompilationMetrics) -> None:
+    If a CompileContext is active, return its compile_id. Otherwise, return the
+    compile_id of the current RuntimeCompileContext.
+    """
+    compile_id = torch._guards.CompileContext.current_compile_id()
+    if compile_id is not None:
+        return compile_id
+    return RuntimeCompileContext.current_compile_id()
```
this makes me think that you should just use CompileContext

masnesral replied:
Do you mean to set the CompileContext at runtime in essentially the same way that I'm setting the RuntimeCompileContext in this PR? If so, then I'd be worried about changing the meaning of what a CompileContext is and when it's available. Currently, there's only a CompileContext available when we're compiling. It seems like there could be future confusion if it's available at runtime too. TLDR I don't know the repercussions.

A modification I could imagine is adding a bit to CompileContext to distinguish between compile-time vs. runtime mode? I'd still be vaguely worried about existing or future callsites asking for a CompileContext and getting back a "runtime" CompileContext...

xmfan replied:
> there's only a CompileContext available when we're compiling

then I think we should assert that there's no active CompileContext, rather than giving it priority over the active RuntimeCompileContext

xmfan replied:
and maybe split this util to make it explicit at the callsite that we are accessing the runtime context

masnesral replied:
So I'm actually using compile_context == None here to determine if we're in runtime vs. compile time. Do you know if there's a more direct way?

xmfan commented on torch/_dynamo/utils.py:
```diff
@@ -1320,6 +1297,60 @@ def add_compilation_metrics_to_chromium(c: CompilationMetrics) -> None:
+
+
+@contextmanager
+def runtime_compile_context(context: Optional[RuntimeCompileContext]):
```
something like this

```suggestion
def runtime_compile_context(context: Optional[RuntimeCompileContext]):
    assert CompileContext.try_get() is None
```

and asserting no runtime compile context from compile_context. This runs in circular import, so we may want to consolidate both into the same file and TLS

masnesral committed:
Update on "[logging] Save compile state in CompiledFxGraph and make it available at runtime"


Summary: To support logging the correct compile_id for runtime timings (like Triton autotuning), save the compile_id in CompiledFxGraph make it available to logging utilities, i.e., dynamo_timed.

The previous attempt put the compile_id in the inductor_metadata with the Triton output code, but that broke Triton caching and we reverted. This version does the following:
* When creating or deserializing a CompiledFxGraph, save the compile-time compile_id.
* Implement a class `RuntimeCompileContext` that's analogous to `CompileContext` where we can look up the compile_id at runtime.
* Set this runtime compile context during `CompiledFxGraph.__call__`.
* Removes the compile_id as a param to dynamo_timed; dynamo_timed can figure it out instead.
* Removes separate dynamo_timed params for compile-time and runtime dynamo_compile column names. We can use one param have dynamo_timed figure out whether to treat as a runtime or compile-time event.

Test Plan:
* tlparse (`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`): https://fburl.com/bu5i8efk
* dynamo_compile: https://fburl.com/scuba/dynamo_compile/sandbox/3d74ps92
* pt2_compile_events: https://fburl.com/scuba/pt2_compile_events/ooqoe5tu

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

xmfan commented on torch/_inductor/output_code.py:
```diff
@@ -451,16 +454,29 @@ def __init__(
+        # happens at runtime. For logging, We still consider that to be "compile time",
+        # so we set the runtime compile context only when a compile context is not set.
+        set_ctx = torch._guards.CompileContext.try_get() is None
+        with runtime_compile_context(self._runtime_ctx) if set_ctx else nullcontext():
```
when does CompiledFxGraph.__call__ get called? every time we cache hit? or only during the initial compilation?

masnesral replied:
CompiledFxGraph is basically doubling as the callable and the entry that's stored in the cache. Though the actual underlying callable (`current_callable`) can't be serialized to disk, so that field is nulled out before serializing and set after a cache hit. 

Not sure if it's helpful, but these are the relevant lines where the callable is set on initial compilation, nulled out before saving to cache, reset after loading from cache:
https://github.com/pytorch/pytorch/blob/579b9f2ed9e2572734b052ea5897354f2e5f74c6/torch/_inductor/output_code.py#L342
https://github.com/pytorch/pytorch/blob/579b9f2ed9e2572734b052ea5897354f2e5f74c6/torch/_inductor/output_code.py#L518
https://github.com/pytorch/pytorch/blob/579b9f2ed9e2572734b052ea5897354f2e5f74c6/torch/_inductor/output_code.py#L554

Anyway, this `__call__` is executed every time the compiled module is actually executed at runtime (the `__call__` part has nothing to do with a cache hit).

xmfan commented:
in terms of unit tests, you could define a custom op, whose real implementation will only be ran at "runtime" and can be used to check if there's an active runtime context. might be worthy to ensure that graph breaks don't have an active context

masnesral committed:
Update on "[logging] Save compile state in CompiledFxGraph and make it available at runtime"


Summary: To support logging the correct compile_id for runtime timings (like Triton autotuning), save the compile_id in CompiledFxGraph make it available to logging utilities, i.e., dynamo_timed.

The previous attempt put the compile_id in the inductor_metadata with the Triton output code, but that broke Triton caching and we reverted. This version does the following:
* When creating or deserializing a CompiledFxGraph, save the compile-time compile_id.
* Implement a class `RuntimeCompileContext` that's analogous to `CompileContext` where we can look up the compile_id at runtime.
* Set this runtime compile context during `CompiledFxGraph.__call__`.
* Removes the compile_id as a param to dynamo_timed; dynamo_timed can figure it out instead.
* Removes separate dynamo_timed params for compile-time and runtime dynamo_compile column names. We can use one param have dynamo_timed figure out whether to treat as a runtime or compile-time event.

Test Plan:
* tlparse (`python benchmarks/dynamo/torchbench.py --performance --training --amp --backend inductor --device cuda --print-compilation-time --repeat 5 --cold-start-latency --only nanogpt`): https://fburl.com/bu5i8efk
* dynamo_compile: https://fburl.com/scuba/dynamo_compile/sandbox/3d74ps92
* pt2_compile_events: https://fburl.com/scuba/pt2_compile_events/ooqoe5tu

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_guards.py:
```diff
@@ -763,6 +763,32 @@ def current_trace_id():
+
+    @staticmethod
+    def try_get() -> Optional[RuntimeCompileContext]:
+        return getattr(_TLS, "runtime_compile_context", None)
```
Would `runtime_compile_context` and `compile_context` ever be set at the same time?  What would that mean?  There also seems like there is overlap with TracingContext below.

If all three of these can be set at the same time, would the trace_id differ?

I'm not totally convinced we need all 3.  Maybe we could have a single context, but with subclasses with more or less information.   That would also simplify lookups.

jansel commented on torch/_inductor/output_code.py:
```diff
@@ -451,16 +454,29 @@ def __init__(
+        # happens at runtime. For logging, We still consider that to be "compile time",
+        # so we set the runtime compile context only when a compile context is not set.
+        set_ctx = torch._guards.CompileContext.try_get() is None
+        with runtime_compile_context(self._runtime_ctx) if set_ctx else nullcontext():
```
I'm concerned about this adding overhead, since this is very much on the hot path and contextlib.contexmanager is slow.  Maybe we could only do it on the first call?

masnesral replied:
@jansel I think there are other runtime events that happen on the 2nd and 3rd calls -- cudagraphy stuff? The actual logging, e.g., of the autotune scuba logs will only happen once. But yeah, I didn't know how concerned to be about the steady-state overhead here. We have:
* The context manager.
* Storing the CompileId in TLS (plus a TLS lookup for an assert)
* The call to `.finish()` in the finally block. (It will do nothing in the steady-state, but still a call)

Can we safely remove the context manager? It's only there to handle a case where we enter this code path twice and the compile-time compileId is different. But is that even possible? Also, we could remove usage of TLS if there's no possibility of multiple threads (but I'd assume there is?). 

Or is there some other place we can stuff away the compileId such that we can look it up at runtime?

jansel replied:
I think all the compiling should be done after the first call.

jansel commented on torch/_inductor/output_code.py:
```diff
@@ -451,16 +454,29 @@ def __init__(
+            try:
+                return self.current_callable(inputs)
+            finally:
+                get_runtime_metrics_context().finish(
```
get_runtime_metrics_context() sounds very similar to the new context you are adding.  Do we really need another one?

masnesral commented:
abandoning this in favor of https://github.com/pytorch/pytorch/pull/148693
